api/src/main/java/com/graphhopper/client/AbstractHopperClient.java

```diff
@@ -0,0 +1,122 @@
+package com.graphhopper.client;
+
+
+import com.graphhopper.GHResponse;
+import com.graphhopper.HopperClient;
+import com.graphhopper.HopperRequest;
+import com.graphhopper.HopperResponse;
+import com.graphhopper.bean.RouteBbox;
+import com.graphhopper.bean.RouteInstruction;
+import com.graphhopper.http.GraphHopperWeb;
+import com.graphhopper.http.WebHelper;
+import com.graphhopper.internal.HopperEngine;
+import com.graphhopper.internal.LazyPointList;
+import com.graphhopper.util.*;
+import com.graphhopper.util.shapes.BBox;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+// Please do not throw an error if hasError (with check everywhere)
+// Throws errors on errors is an anti-pattern
+public abstract class AbstractHopperClient implements HopperClient {
+
+    // All child-classes must provide a customized engine
+    public abstract HopperEngine getEngine();
+
+    @Override
+    public HopperResponse route(HopperRequest request) {
+        /**
+         * This method is a mess, and I cheated alot here using the downloader.
+         * For real the method must take in consideration the "fetchXXX" from HopperRequest
+         * and use the HopperEngine to generate the info and populate the beans.
+         *
+         * Note that the refactor must GO DEEPER (using the beans) for real performance improvements.
+         *
+         * DEPP REFACTOR TODO:
+         * - com.graphhopper.util.PathMerger (can be merged inside here, actually)
+         * - com.graphhopper.routing.Path
+         */
+        Downloader downloader = new Downloader("GraphHopper Test") {
+            @Override
+            public InputStream fetch( String url ) throws IOException {
+                return getClass().getClassLoader().getResourceAsStream("com.graphhopper/test.json");
+            }
+        };
+
+        GraphHopperWeb instance = new GraphHopperWeb().setPointsEncoded(false);
```

Comment:
AbstractHopperClient right now is just a mock, read the roadmap.  TODO:  Working AbstractHopperClient.route() populating beans without instancing XInstruction InstructionLists or PointLists at all (aka PathMerger and Path logic, but with beans)